src/sabre_swap/layer.rs

```diff
@@ -0,0 +1,295 @@
+    /// front layer is important.  We need to maintain the insertion order even with removals from
+    /// the layer.
+    iteration_order: Vec<Option<NodeIndex>>,
+    /// The index of the first populated entry in the `iteration_order`.  If the iteration order is
+    /// empty, this will be 0.
+    iteration_start: usize,
+    /// The index one past the last populated entry in the `iteration_order`.  If the iteration
+    /// order is empty, this will be 0.
+    iteration_end: usize,
+}
+
+impl FrontLayer {
+    pub fn new(num_qubits: usize) -> Self {
+        FrontLayer {
+            // This is the maximum capacity of the front layer, since each qubit must be one of a
+            // pair, and can only have one gate in the layer.
+            nodes: HashMap::with_capacity(num_qubits / 2),
+            qubits: vec![None; num_qubits],
+            iteration_order: vec![None; num_qubits],
+            iteration_start: 0,
+            iteration_end: 0,
+        }
+    }
+
+    /// Add a node into the front layer, with the two qubits it operates on.  This usually has
+    /// constant-time complexity, except if the iteration-order buffer is full.
+    pub fn insert(&mut self, index: NodeIndex, qubits: [usize; 2]) {
+        let [a, b] = qubits;
+        self.qubits[a] = Some((index, b));
+        self.qubits[b] = Some((index, a));
+        self.nodes.insert(index, qubits);
+
+        self.iteration_order[self.iteration_end] = Some(index);
+        self.iteration_end += 1;
+        if self.iteration_end == self.iteration_order.len() {
+            // Condense items back to the start of the vector.
+            let mut ptr = 0;
+            for i in self.iteration_start..self.iteration_end {
+                if let Some(value) = self.iteration_order[i] {
+                    self.iteration_order[i] = None;
+                    self.iteration_order[ptr] = Some(value);
+                    ptr += 1;
+                }
+            }
+            self.iteration_start = 0;
+            self.iteration_end = ptr;
+        }
+    }
+
+    /// Remove a node from the front layer.
+    pub fn remove(&mut self, index: &NodeIndex) {
+        let [q0, q1] = self.nodes.remove(index).unwrap();
+        self.qubits[q0] = None;
+        self.qubits[q1] = None;
+
+        // If the element was at the start of the iteration order, advance the pointer.
+        match self.iteration_order[self.iteration_start] {
+            Some(a) if a == *index => {
+                self.iteration_order[self.iteration_start] = None;
+                if self.iteration_start + 1 == self.iteration_end {
+                    self.iteration_start = 0;
+                    self.iteration_end = 0;
+                }
+                while self.iteration_start < self.iteration_end
+                    && self.iteration_order[self.iteration_start].is_none()
+                {
+                    self.iteration_start += 1;
+                }
```

Comment:
A really tiny nit micro-optimization benchmarking question here which we should just disregard it really as it's more idle musing, especially as I know this block will be removed in the subsequent PR This just looked a bit odd to me as a way to find the index of the first non-None element in a slice. It works fine, but I wonder how it would compare performance wise to something like (untested so there are probably typos or mistakes):

```rust
self.iteration_start = match self.iteration_order[self.iteration_start..].iter().position(|x| x.is_some()) {
    Some(relative_index) => relative_index + self.iteration_start,
    None => self.iteration_end,
};
```

I expect this will be slower in the non-match case as it will traverse the full vec instead of terminating at insertion_end (although you could probably work around that with `enumerate()` and some other logic. It really is a question of how much the compiler can optimize with an iterator instead of a while loop I guess.

Reply:
The answer to most questions about why I used loops and not iterators is going to be "because I didn't know there was an iterator method that does it" haha.  I think in your example, we could just limit the slice to `self.iteration_start..self.iteration_end` to get the early-exit behaviour too, though?